pkg/controller/kieapp/defaults/defaults.go

```diff
@@ -108,46 +108,75 @@ func getConsoleTemplate(cr *v1.KieApp) v1.ConsoleTemplate {
 	}
-	names := map[string]bool{}
+	usedNames := map[string]bool{}
 	for index := range cr.Spec.Objects.Servers {
-		kieID := GenKieName(cr.Name, cr.Spec.Objects.Servers[index].Name)
-		if exists := names[kieID]; exists {
-			return []v1.ServerTemplate{}, fmt.Errorf("duplicate kieserver name %s", kieID)
+		serverSet := &cr.Spec.Objects.Servers[index]
+		if serverSet.Name == "" {
+			if len(cr.Spec.Objects.Servers) == 1 {
+				serverSet.Name = fmt.Sprintf("%v-kieserver", cr.Name)
+			} else {
+				serverSet.Name = fmt.Sprintf("%v-kieserver%v", cr.Name, index)
+			}
+		} else if usedNames[serverSet.Name] {
+			return []v1.ServerTemplate{}, fmt.Errorf("duplicate kieserver name %s", serverSet.Name)
+		} else {
+			usedNames[serverSet.Name] = true
 		}
-		if cr.Spec.Objects.Servers[index].Deployments == nil {
-			cr.Spec.Objects.Servers[index].Deployments = Pint(constants.DefaultKieDeployments)
+		if serverSet.Deployments == nil {
+			serverSet.Deployments = Pint(constants.DefaultKieDeployments)
 		}
-		for i := 0; i < *cr.Spec.Objects.Servers[index].Deployments; i++ {
-			names[kieID] = true
+		crTemplate := v1.ServerTemplate{
+			Build: getBuildConfig(commonConfig, serverSet.Build),
 		}
-	}
-	for index := range cr.Spec.Objects.Servers {
-		for kieID := range names {
-			if kieID == GenKieName(cr.Name, cr.Spec.Objects.Servers[index].Name) {
-				for i := 0; i < *cr.Spec.Objects.Servers[index].Deployments; i++ {
-					crTemplate := v1.ServerTemplate{KieName: kieID, KieIndex: i}
-					crTemplate.Build = getBuildConfig(commonConfig, cr.Spec.Objects.Servers[index].Build)
-					if cr.Spec.Objects.Servers[index].Build != nil {
-						crTemplate.From = getKieServerImageForBuild(commonConfig, i)
-					} else {
-						crTemplate.From = getDefaultKieServerImage(commonConfig, cr.Spec.Objects.Servers[index].From)
-					}
-					servers = append(servers, crTemplate)
-				}
+		if serverSet.Build != nil {
+			if *serverSet.Deployments > 1 {
+				return []v1.ServerTemplate{}, fmt.Errorf("Cannot request %v deployments for a build", *serverSet.Deployments)
+			}
+			crTemplate.From = corev1.ObjectReference{
+				Kind:      "ImageStreamTag",
+				Name:      fmt.Sprintf("%s-kieserver:latest", commonConfig.ApplicationName),
+				Namespace: "",
+			}
+		} else {
+			crTemplate.From = getDefaultKieServerImage(commonConfig, serverSet.From)
+		}
+		crTemplate.KieName = serverSet.Name
+		if *serverSet.Deployments == 1 {
+			crTemplate.KieIndex = GetKieIndex(serverSet, 0)
+			servers = append(servers, crTemplate)
+		} else {
+			for i := 0; i < *serverSet.Deployments; i++ {
+				instanceTemplate := crTemplate.DeepCopy()
+				instanceTemplate.KieIndex = GetKieIndex(serverSet, i)
+				servers = append(servers, *instanceTemplate)
 			}
 		}
 	}
 	return servers, nil
 }
 
-// GenKieName generates the KieName for a given server
-func GenKieName(crName, serverName string) string {
-	if serverName != "" {
-		return strings.Join([]string{crName, serverName}, "-")
```

Comment:
This should stay, it ensure a unique kieserver name within a given namespace by prepending the cr name to the front a given serverset name. Without it, multi-tenant deployments can/will get very messy and conflict.

Reply:
We have to give the user control to choose the names they want. If they select "credit-card-statement" or whatever, that should be their exact service name. Prepending it would prevent that. I think multiple environments in one namespace is unlikely, and if we were gonna go down the road of anticipating that, might as well worry about non-RHPAM components with conflicting name.

I think instead, we should separately do a PR (later) to pre-emptively validate the CR and fail with a clear message if it results in duplicate components.